src/pyramid/request.py

```diff
@@ -330,3 +333,117 @@ def apply_request_extensions(request, extensions=None):
         InstancePropertyHelper.apply_properties(
             request, extensions.descriptors
         )
+
+
+class RequestLocalCache:
+    """
+    A store that caches values during for the lifecycle of a request.
+
+    Instantiate a cache object and use it to decorate methods or functions
+    that accept a request parameter. Any other arguments are not used as
+    cache keys so make sure they are constant across calls per-request.
+
+    Wrapping methods:
+
+    .. code-block:: python
+
+        class SecurityPolicy:
+            identity_cache = RequestLocalCache()
+
+            @identity_cache
+            def authenticated_identity(self, request):
+                result = ...  # do some expensive computations
+                return result
+
+    Wrapping functions:
+
+    .. code-block:: python
+
+        user_cache = RequestLocalCache()
+
+        @user_cache
+        def get_user(request):
+            result = ...  # do some expensive computations
+            return result
+
+    It's also possible to inspect and influence the cache during runtime using
+    :meth:`.get`, :meth:`.set` and :meth:`.clear`. Using these methods, the
+    cache can be used directly as well, without using it as a decorator.
+
+    The cache will release resources aggressively by utilizing
+    :meth:`pyramid.request.Request.add_finished_callback`, but it will also
+    maintain a weakref to the request and cleanup when it is garbage collected
+    if the callbacks are not invoked for some reason.
+
+    .. versionadded:: 2.0
+
+    """
+
+    NO_VALUE = Sentinel('NO_VALUE')
+
+    def __init__(self):
+        self.store = weakref.WeakKeyDictionary()
+
+    def auto_adapt(decorator):
+        class FunctionOrMethodAdapter:
+            def __init__(self, inst, func):
+                self.inst = inst
+                self.__wrapped__ = func
+
+            def __call__(self, *args, **kwargs):
+                return decorator(self.inst, self.__wrapped__)(*args, **kwargs)
+
+            def __get__(self, instance, owner):
+                return decorator(
+                    self.inst, self.__wrapped__.__get__(instance, owner)
+                )
+
+        def adapt(inst, fn):
+            return FunctionOrMethodAdapter(inst, fn)
+
+        return adapt
+
+    @auto_adapt
+    def __call__(self, fn):
+        """
+        Decorate a method or function.
+
+        """
+
+        @functools.wraps(fn)
```

Comment:
This whole auto_adapt thing negates `wraps`, it swallows `__name__`, `__doc__`, et al.

```python
from pyramid.request import RequestLocalCache


cache = RequestLocalCache()


@cache
def myfunc(request):
    """
    Return hello.

    """
    return 'hello'


print(myfunc.__name__)  # should print "myfunc", but raises AttributeError
print(myfunc.__doc__)  # should print "Return hello", but prints "None"
```

Reply:
Yeah I couldn't come up with an approach that preserved that stuff correctly unfortunately. The cost of doing business unless someone can come up with a better approach. I looked at using the `wrapt` package initially and it returned things like `FunctionWrapper` objects, etc so I decided this wasn't really worse than that.

I'm comfortable with the current approach and am pretty certain we can fix it in the future without compatibility concerns so unless you actually know what to do, I'm gonna leave it.

Reply:
Ok I've pushed a new API for this and got rid of the original decorator approach. The original approach suffered from some issues like sharing the cache object across instances of the policy when used at the class level, and was error-prone in that it accepted extra arguments into the function. Now it only accepts a request, always, and it's happy.